sympy/matrices/matrices.py

```diff
@@ -3912,35 +3975,121 @@ def LUdecomposition_Simple(self, iszerofunc=_iszero):
         LUdecompositionFF
         LUsolve
         """
-        if not self.is_square:
-            raise NonSquareMatrixError(
-                "A Matrix must be square to apply LUdecomposition_Simple().")
-        n = self.rows
-        A = self.as_mutable()
-        p = []
-        # factorization
-        for j in range(n):
-            for i in range(j):
-                for k in range(i):
-                    A[i, j] = A[i, j] - A[i, k] * A[k, j]
-            pivot = -1
-            for i in range(j, n):
-                for k in range(j):
-                    A[i, j] = A[i, j] - A[i, k] * A[k, j]
-                # find the first non-zero pivot, includes any expression
-                if pivot == -1 and not iszerofunc(A[i, j]):
-                    pivot = i
-            if pivot < 0:
-                # this result is based on iszerofunc's analysis of the possible pivots, so even though
-                # the element may not be strictly zero, the supplied iszerofunc's evaluation gave True
-                raise ValueError("No nonzero pivot found; inversion failed.")
-            if pivot != j:  # row must be swapped
-                A.row_swap(pivot, j)
-                p.append([pivot, j])
-            scale = 1 / A[j, j]
-            for i in range(j + 1, n):
-                A[i, j] = A[i, j] * scale
-        return A, p
+
+        if rankcheck:
+            SymPyDeprecationWarning(
+                feature="Keyword argument rankcheck is deprecated. "
+                        "Future LUdecomposition_Simple() will complete "
+                        "regardless rank of input matrix.",
+                issue=9796,
+                deprecated_since_version="1.1").warn()
+
+        if self.rows == 0 or self.cols == 0:
+            # Define LU decomposition of a matrix with no entries as a matrix
+            # of the same dimensions with all zero entries.
+            return self.zeros(self.rows, self.cols), []
+
+        lu = self.as_mutable()
+        row_swaps = []
+
+        pivot_col = 0
+        for pivot_row in range(0, lu.rows-1):
+            # Search for pivot. Prefer entry that iszeropivot determines
+            # is nonzero, over entry that iszeropivot cannot guarantee
+            # is  zero.
+            candidate_pivot_row = None
+            while pivot_col != self.cols:
+                for r in range(pivot_row, self.rows):
+                    iszeropivot = iszerofunc(lu[r, pivot_col])
```

Comment:
I opted for the current implementation because _find_reasonable_pivot introduced a fairly significant slowdown for symbolic matrices, but I may be doing it wrong. To test _find_reasonable_pivot's performance, I added a call to it just before the existing pivot search in LUdecomposition_Simple(), and ignored its output:

```python
            while pivot_col != self.cols:
                for r in range(pivot_row, self.rows):

                    pivot_offset, pivot_val, assumed_nonzero, newly_determined =\
                        _find_reasonable_pivot(lu[pivot_row:, pivot_col],
                                               iszerofunc=iszerofunc,
                                               simpfunc=_simplify)

                    iszeropivot = iszerofunc(lu[r, pivot_col])
                    if iszeropivot is None and candidate_pivot_row is None:
                        # iszerofunc could not determine if
                        # lu[r, pivot_col] is zero.
                        # Save this row in case we don't
                        # find a nonzero entry during the search
                        # of this column
                        candidate_pivot_row = r
                    elif iszeropivot == False:
                        # iszerofunc determined that
                        # lu[r, pivot_col] is zero.
                        # Use this entry as the pivot
                        candidate_pivot_row = r
                        break
                    elif iszeropivot == True:
                        lu[r, pivot_col] = S.Zero
                if candidate_pivot_row is None:
                    # There is no nonzero pivot in this column.
                    # Exited search within column because
                    # iszeropivot() returned False for each
                    # entry
                    pivot_col += 1
                else:
                    # lu[r, pivot_col] is either non-zero,
                    # or iszeropivot could not determine that it is
                    # zero.
                    break
```
I then ran the following script twice with the call to _find_reasonable_pivot() commented out in LUdecomposition_Simple() on the second run.
```sympy
import sympy

def main():

    lu, perm = sympy.Matrix(4, 2, map(sympy.Symbol, "abcdefgh")).LUdecomposition_Simple()

    sympy.pretty_print(lu)
    print perm

if __name__ == "__main__":

    for k in range(10):
        main()
```
The profiler results of both runs are:
```
>>> p = pstats.Stats('/tmp/find_reasonable_pivot_on')
>>> p.sort_stats('cumulative').print_stats(10)
Wed Mar 29 21:06:21 2017    /tmp/find_reasonable_pivot_on

         21332950 function calls (20699591 primitive calls) in 55.268 seconds

   Ordered by: cumulative time
   List reduced from 2574 to 10 due to restriction <10>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000   55.465   55.465 find_reasonable_pivot_test.py:1(<module>)
       10    0.000    0.000   53.589    5.359 find_reasonable_pivot_test.py:3(main)
       10    0.002    0.000   53.279    5.328 sympy/matrices/matrices.py:3930(LUdecomposition_Simple)
       70    0.005    0.000   53.149    0.759 sympy/matrices/matrices.py:5269(_find_reasonable_pivot)
  430/250    0.018    0.000   44.144    0.177 sympy/core/expr.py:595(equals)
       90    0.005    0.000   35.123    0.390 sympy/core/expr.py:440(is_constant)
      180    0.012    0.000   29.239    0.162 sympy/core/expr.py:362(_random)
      720    0.012    0.000   28.823    0.040 sympy/utilities/randtest.py:16(random_complex_number)
1800/1440    0.027    0.000   28.763    0.020 sympy/simplify/simplify.py:1159(nsimplify)
      360    0.029    0.000   28.669    0.080 sympy/simplify/simplify.py:1284(_real_to_rational)


<pstats.Stats instance at 0x10c6e9248>
>>> p = pstats.Stats('/tmp/find_reasonable_pivot_off')
>>> p.sort_stats('cumulative').print_stats(10)
Wed Mar 29 21:06:40 2017    /tmp/find_reasonable_pivot_off

         373170 function calls (366705 primitive calls) in 1.418 seconds

   Ordered by: cumulative time
   List reduced from 2108 to 10 due to restriction <10>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.001    0.001    1.424    1.424 find_reasonable_pivot_test.py:1(<module>)
        1    0.005    0.005    0.844    0.844 sympy/__init__.py:12(<module>)
       10    0.000    0.000    0.579    0.058 find_reasonable_pivot_test.py:3(main)
       10    0.000    0.000    0.448    0.045 sympy/printing/pretty/pretty.py:2162(pretty_print)
       10    0.000    0.000    0.430    0.043 sympy/printing/pretty/pretty.py:2144(pretty)
       10    0.000    0.000    0.429    0.043 sympy/printing/pretty/pretty.py:61(doprint)
   449/10    0.006    0.000    0.423    0.042 sympy/printing/printer.py:235(_print)
       10    0.000    0.000    0.423    0.042 sympy/printing/pretty/pretty.py:664(_print_MatrixBase)
       10    0.003    0.000    0.419    0.042 sympy/printing/pretty/pretty.py:595(_print_matrix_contents)
        1    0.005    0.005    0.331    0.331 sympy/polys/__init__.py:1(<module>)


<pstats.Stats instance at 0x10b011b48>
```